test/StreamJsonRpc.Tests/JsonContractResolverTests.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.VisualStudio.Threading;
+using Nerdbank.Streams;
+using Newtonsoft.Json.Serialization;
+using StreamJsonRpc;
+using Xunit;
+using Xunit.Abstractions;
+
+public partial class JsonContractResolverTest : TestBase
+{
+    protected readonly Server server = new Server();
+    protected readonly JsonRpc serverRpc;
+    protected readonly JsonRpc clientRpc;
+    protected readonly IServer client;
+
+    private const string ExceptionMessage = "Some exception";
+
+    public JsonContractResolverTest(ITestOutputHelper logger)
+        : base(logger)
+    {
+        var pipes = FullDuplexStream.CreatePipePair();
+
+        this.client = JsonRpc.Attach<IServer>(new LengthHeaderMessageHandler(pipes.Item1, this.CreateFormatter()));
+        this.clientRpc = ((IJsonRpcClientProxy)this.client).JsonRpc;
+
+        this.serverRpc = new JsonRpc(new LengthHeaderMessageHandler(pipes.Item2, this.CreateFormatter()));
+        this.serverRpc.AddLocalRpcTarget(this.server);
+
+        this.serverRpc.TraceSource = new TraceSource("Server", SourceLevels.Verbose);
+        this.clientRpc.TraceSource = new TraceSource("Client", SourceLevels.Verbose);
+
+        this.serverRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+        this.clientRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+
+        this.serverRpc.StartListening();
+    }
+
+    protected interface IServer : IDisposable
+    {
+        Task PushCompleteAndReturn(IObserver<int> observer);
+    }
+
+    [Fact]
+    public async Task PushCompleteAndReturn()
+    {
+        var observer = new MockObserver<int>();
+        await Task.Run(() => this.client.PushCompleteAndReturn(observer)).WithCancellation(this.TimeoutToken);
+        await observer.Completion.WithCancellation(this.TimeoutToken);
+        ImmutableList<int> result = await observer.Completion;
+    }
+
+    protected IJsonRpcMessageFormatter CreateFormatter()
+    {
+        var formatter = new JsonMessageFormatter();
+        formatter.JsonSerializer.ContractResolver = new CamelCasePropertyNamesContractResolver();
```

Comment:
I feel like adding a test that combines your new marshaled interfaces feature with use of `IObserver<T>` and a custom contract resolver would go a long way to alleviating concern that these will all work together.

Reply:
I'd like to see the same formatter work for both IObserver and custom interface marshaling, while configured with a custom contract resolver.